core/templates/App.ts

```diff
@@ -274,6 +274,26 @@ angular.module('oppia').factory('$exceptionHandler', [
       if (UNHANDLED_REJECTION_STATUS_CODE_REGEX.test(exception)) {
         return;
       }
+      // If an unhandled rejection is encountered in the AngularJS promise
+      // manager, it checks the data type of the rejection value. If it
+      // satifies the conditions below, it calls the exceptionHandler with the
+      // error object as the first argument and an message string as the
+      // second argument.
+      // 1. It is an instance of Error type.
+      // 2. It is an Error Object.
+      // 3. It is Exception Object.
+      // 4. It is a DOMException Object.
+      // eslint-disable-next-line max-len
+      // see: https://github.com/angular/angular.js/blob/2dfb6b4af62d750032c91fd86dc1f8d684d179c6/src/Angular.js#L628
+      // eslint-disable-next-line max-len
+      // see: https://github.com/angular/angular.js/blob/2dfb6b4af62d750032c91fd86dc1f8d684d179c6/src/ng/q.js#L388
+      // Otherwise, it will call the exceptionHandler with only the error
+      // message. In order to log the error correctly on StackDriver and to
+      // preserve the original stacktrace, we wrap such exceptions in an error
+      // object.
+      if (exception.message === undefined) {
```

Comment:
Shouldn't you be checking the type of the object? E.g. if exception instanceof String, or similar?

Also how does this preserve the original stacktrace?

Perhaps worth adding frontend tests for all these cases, too.

Reply:
Added some code to do the same type checking angular does internally. Added tests for the exceptionHandler as well as the isError type-check.

When an error object is created (or thrown in case of older browsers), the stack property is updated with the stack trace until that point. I've verified this for angularJS unhandled promise rejection errors. Here's a screenshot of such a stacktrace sent to the backend:

![image](https://user-images.githubusercontent.com/11008603/82677745-1fa92f00-9c66-11ea-9c9d-8c264798e8e3.png)

Note that it is clear that the error originated from angular and that the error message is "Possibly unhandled rejection: backdrop click"